### PR TITLE
Feature/queue existence check for provider

### LINKED ISSUE #45: Check  if queue exists before sending the message
## New Feature

On producer, check if a queue exists before sending the message and reject produce function if it does not not.

Currently:

``` js
 producer.produce('non existent queue',  { msg: Date.now() }, {rpc:true}).then(...)
```

Waits indefinitely or resolves immediately in case `rpc: false` for a queue that does not exists.
## Acceptance

Check if the queue exists before sending the message and if not, reject the produce function with an appropriate error denoting that the queue does not exist.

### PR DESCRIPTION
**Pull request by mrister**
In a provider reject the produce method in case there is no queue. Closes #45

### CODE REVIEW

File: src/modules/producer.js [hunk 1]
```diff
@@ -50,28 +69,31 @@ function createRpcQueue(queue) {
...
   };
 }
 
 function publishOrSendToQueue(queue, msg, options) {
-  if (!options.routingKey) {
-    return this.channel.sendToQueue(queue, msg, options);
-  } else {
-    return this.channel.publish(queue, options.routingKey, msg, options);
-  }
+  return checkQueue(this.channel, queue)
```
Comment (PuKoren) [inline src/modules/producer.js:105]: This will query the broker each time we want to send a message?

File: src/modules/producer.js [hunk 2]
```diff
@@ -13,33 +28,43 @@ function createRpcQueue(queue) {
   if (!amqpRPCQueues[queue]) {
     amqpRPCQueues[queue] = {};
   }
-
-  let rpcQueue = amqpRPCQueues[queue];
+  
+  const rpcQueue = amqpRPCQueues[queue];
   if (rpcQueue.queue) return Promise.resolve(rpcQueue.queue);
-
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 3]
```diff
@@ -13,33 +28,43 @@ function createRpcQueue(queue) {
...
-    .catch(() => {
+    .catch((err) => {
       delete rpcQueue.queue;
+      if (err.type === 'no_queue') throw err;
       return utils.timeoutPromise(this.conn.config.timeout).then(() => {
         return createRpcQueue.call(this, queue);
       });
     });
-
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 4]
```diff
@@ -81,27 +109,28 @@ function publishOrSendToQueue(queue, msg, options) {
...
+        return publishOrSendToQueue.call(this, queue, msg, options)
+          .then(() => {
+            //defered promise that will resolve when response is received
+            amqpRPCQueues[queue][corrId] = Promise.defer();
+            return amqpRPCQueues[queue][corrId].promise;
+          });
+      });
   }
-
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 5]
```diff
@@ -50,28 +75,31 @@ function createRpcQueue(queue) {
  */
 function maybeAnswer(queue) {
   var rpcQueue = amqpRPCQueues[queue];
-
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 6]
```diff
@@ -13,33 +28,43 @@ function createRpcQueue(queue) {
...
+  rpcQueue.queue = this
+    .conn
+    .get()
+    .then((channel) => {
+      return checkQueue(channel, resQueue)
+        .then(()=> {
+          return channel.assertQueue(resQueue, {durable: true, exclusive: true})
+            .then((_queue) => {
+              rpcQueue.queue = _queue.queue;
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 7]
```diff
@@ -50,28 +75,31 @@ function createRpcQueue(queue) {
  */
 function maybeAnswer(queue) {
   var rpcQueue = amqpRPCQueues[queue];
-
+  
   return (msg) => {
     //check the correlation ID sent by the initial message using RPC
     var corrId = msg.properties.correlationId;
-
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 8]
```diff
@@ -81,27 +109,28 @@ function publishOrSendToQueue(queue, msg, options) {
  * @param  {object} options contain rpc property (if true, enable rpc for this message)
  * @return {Promise}         Resolves when message is correctly sent, or when response is received when rpc is enabled
  */
-function checkRpc (queue, msg, options) {
+function checkRpc(queue, msg, options) {
   //messages are persistent
   options.persistent = true;
-
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 9]
```diff
@@ -115,29 +144,30 @@ function checkRpc (queue, msg, options) {
 function produce(queue, msg, options) {
   //default options are persistent and durable because we do not want to miss any outgoing message
   //unless user specify it
-  options = Object.assign({ persistent: true, durable: true }, options);
-
+  options = Object.assign({persistent: true, durable: true}, options);
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 10]
```diff
@@ -13,33 +28,43 @@ function createRpcQueue(queue) {
...
+          return channel.assertQueue(resQueue, {durable: true, exclusive: true})
+            .then((_queue) => {
+              rpcQueue.queue = _queue.queue;
+              
+              //if channel is closed, we want to make sure we cleanup the queue so future calls will recreate it
+              this.conn.addListener('close', () => {
+                delete rpcQueue.queue;
+                createRpcQueue.call(this, queue);
+              });
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 11]
```diff
@@ -115,29 +144,30 @@ function checkRpc (queue, msg, options) {
...
-  })
-  .catch((err) => {
-    //add timeout between retries because we don't want to overflow the CPU
-    this.conn.config.transport.error('bmq:producer', err);
-    return utils.timeoutPromise(this.conn.config.timeout)
-    .then(() => {
-      return produce.call(this, queue, msg, options);
+    .then((_channel) => {
+      this.channel = _channel;
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 12]
```diff
@@ -115,29 +144,30 @@ function checkRpc (queue, msg, options) {
...
-    this.conn.config.transport.error('bmq:producer', err);
-    return utils.timeoutPromise(this.conn.config.timeout)
-    .then(() => {
-      return produce.call(this, queue, msg, options);
+    .then((_channel) => {
+      this.channel = _channel;
+      
+      //undefined can't be serialized/buffered :p
+      if (!msg) msg = null;
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 13]
```diff
@@ -81,27 +109,28 @@ function publishOrSendToQueue(queue, msg, options) {
...
-      amqpRPCQueues[queue][corrId] = Promise.defer();
-      return amqpRPCQueues[queue][corrId].promise;
-    });
+      .then(() => {
+        //generates a correlationId (random uuid) so we know which callback to execute on received response
+        var corrId = uuid.v4();
+        options.correlationId = corrId;
+        //reply to us if you receive this message!
+        options.replyTo = amqpRPCQueues[queue].queue;
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 14]
```diff
@@ -115,29 +144,30 @@ function checkRpc (queue, msg, options) {
...
-    .then(() => {
-      return produce.call(this, queue, msg, options);
+    .then((_channel) => {
+      this.channel = _channel;
+      
+      //undefined can't be serialized/buffered :p
+      if (!msg) msg = null;
+      
+      this.conn.config.transport.info('bmq:producer', '[' + queue + '] > ', msg);
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js [hunk 15]
```diff
@@ -13,33 +28,43 @@ function createRpcQueue(queue) {
   if (!amqpRPCQueues[queue]) {
     amqpRPCQueues[queue] = {};
   }
-
-  let rpcQueue = amqpRPCQueues[queue];
+
```
Comment (dialonce) [inline src/modules/producer.js]: MAJOR Remove the useless trailing whitespaces at the end of this line. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3ATrailingWhitespace)

File: src/modules/producer.js (hunk 1 above)
Comment (mrister) [inline src/modules/producer.js:105]: Hm possibly yes, let me investigate.

**Comment by dialonce** [comment]
SonarQube analysis reported 15 issues
* MAJOR 15 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [producer.js#L27](https://github.com/dial-once/node-bunnymq/blob/1de6d7010a69bd8f661e3cfea3303d7db99e762c/src/modules/producer.js#L27): Function has a complexity of 13 which is greater than 10 authorized. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: src/modules/producer.js [hunk 16]
```diff
@@ -13,33 +28,43 @@ function createRpcQueue(queue) {
...
+              
+              return channel.consume(_queue.queue, maybeAnswer.call(this, queue), {noAck: true});
+            })
+            .then(() => rpcQueue.queue);
+        });
     })
-    .catch(() => {
+    .catch((err) => {
       delete rpcQueue.queue;
+      if (err.type === 'no_queue') throw err;
```
Comment (PuKoren) [inline src/modules/producer.js:72]: This and `checkQueue` are the only changes? I have some difficulty to read the diff as I think your editor reindented all the code in a rampage :laughing:

**Comment by dialonce** [comment]
SonarQube analysis reported 15 issues
* MAJOR 15 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [producer.js#L27](https://github.com/dial-once/node-bunnymq/blob/6be2416f0988fa23639d4809f8b9e8f8a3689b47/src/modules/producer.js#L27): Function has a complexity of 13 which is greater than 10 authorized. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: src/modules/producer.js (hunk 16 above)
Comment (mrister) [inline src/modules/producer.js:72]: Huh, yes, and the tests I'll try to revert the editor changes.

File: src/modules/producer.js (hunk 1 above)
Comment (mrister) [inline src/modules/producer.js:105]: @PuKoren  so if I try this:

``` js
 return Promise.all ([
        producer.produce(fixtures.queues[0],  { msg: Date.now() }, {rpc: true}),
        producer.produce(fixtures.queues[0],  { msg: Date.now() }, {rpc: true}),
        producer.produce(fixtures.queues[0],  { msg: Date.now() }, {rpc: true}),

        producer.produce(fixtures.queues[1],  { msg: Date.now() }, {rpc: true}),
        producer.produce(fixtures.queues[1],  { msg: Date.now() }, {rpc: true})
      ]
```

I see check being performed only once per queue.

Comment (PuKoren) [inline src/modules/producer.js:105]: @mrister how is that possible? It means that amqp.node caches the existance of the queue?
Do your queues exists on your sample code?

Comment (mrister) [inline src/modules/producer.js:105]: They do, when I run the tests using fixtures. I've added the above code in producer-consumer-spec.js for my testing to answer your question.

**Comment by dialonce** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [producer.js#L27](https://github.com/dial-once/node-bunnymq/blob/2f965faec2cefb159a545d1cb8f56f103a620086/src/modules/producer.js#L27): Function has a complexity of 13 which is greater than 10 authorized. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: src/modules/producer.js (hunk 1 above)
Comment (PuKoren) [inline src/modules/producer.js:105]: Ok so `checkQueue` resolves immediately when queue is already opened? Looks good

**Comment by dialonce** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [producer.js#L27](https://github.com/dial-once/node-bunnymq/blob/a20bddc8a797e67d9a4373369d921629a5de2005/src/modules/producer.js#L27): Function has a complexity of 13 which is greater than 10 authorized. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3AFunctionComplexity)

**Comment by dialonce** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [producer.js#L27](https://github.com/dial-once/node-bunnymq/blob/ad8031f0c67765c5454bbc0dbca5af23251804db/src/modules/producer.js#L27): Function has a complexity of 13 which is greater than 10 authorized. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3AFunctionComplexity)

**Comment by dialonce** [comment]
SonarQube analysis reported 1 issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [producer.js#L27](https://github.com/dial-once/node-bunnymq/blob/6a77820d1f2ac22e18b23d4635868d0aeeb7b7d7/src/modules/producer.js#L27): Function has a complexity of 13 which is greater than 10 authorized. [rule](http://sonar.dialonce.io/coding_rules#rule_key=javascript%3AFunctionComplexity)

File: src/modules/producer.js (hunk 1 above)
Comment (mrister) [inline src/modules/producer.js:105]: yup!

File: src/modules/producer.js (hunk 16 above)
Comment (mrister) [inline src/modules/producer.js:72]: should be more readable now.

**Comment by PuKoren** [comment]
LGTM

**Comment by jkernech** [comment]
LGTM

**Pull request merged**